Index: .idea/robot_car2.iml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/robot_car2.iml b/.idea/robot_car2.iml
new file mode 100644
--- /dev/null	(date 1609079602259)
+++ b/.idea/robot_car2.iml	(date 1609079602259)
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<module type="PYTHON_MODULE" version="4">
+  <component name="NewModuleRootManager">
+    <content url="file://$MODULE_DIR$" />
+    <orderEntry type="inheritedJdk" />
+    <orderEntry type="sourceFolder" forTests="false" />
+  </component>
+  <component name="PackageRequirementsSettings">
+    <option name="requirementsPath" value="" />
+  </component>
+</module>
\ No newline at end of file
Index: arduino/frontServo/frontServo.ino
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.BaseRevisionTextPatchEP
<+>#include <Servo.h>\nServo servo;\nint angle = 0;\nvoid setup() {\n  Serial.begin(115200);\n  servo.attach(9); // D9\n//  for (angle = servo.read(); angle > 0; angle--) {\n    servo.write(angle);\n   //delay(1000);\n//  }\n}\nvoid loop() \n{ \n  long ct = millis();\n  \n // scan from 0 to 180 degrees\n  Serial.println(\"0\");\n  servo.write(0);               \n  for(angle = 0; angle < 250; angle=angle+2)  \n  {                                  \n    servo.write(angle);               \n    delay(20);                   \n  }\n  long nt = millis();\n  Serial.println(\"hej\");\n  Serial.println(String(nt-ct));\n \n  // now scan back from 180 to 0 degrees\n  for(angle = 180; angle > 0; angle=angle-2)    \n  {                                \n    servo.write(angle);           \n    delay(20);       \n  }\n  nt = millis();\n  Serial.println(\"hej\");\n  Serial.println(String(nt-ct));\n \n}\n
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/arduino/frontServo/frontServo.ino b/arduino/frontServo/frontServo.ino
--- a/arduino/frontServo/frontServo.ino	(revision 8aae5515d02d115521535e2fcabe18b7b0a6ef5b)
+++ b/arduino/frontServo/frontServo.ino	(date 1607236951363)
@@ -3,7 +3,7 @@
 int angle = 0;
 void setup() {
   Serial.begin(115200);
-  servo.attach(9); // D9
+  servo.attach(3); // D9
 //  for (angle = servo.read(); angle > 0; angle--) {
     servo.write(angle);
    //delay(1000);
Index: .idea/misc.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/misc.xml b/.idea/misc.xml
new file mode 100644
--- /dev/null	(date 1603656109302)
+++ b/.idea/misc.xml	(date 1603656109302)
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectRootManager" version="2" project-jdk-name="Python 3.8" project-jdk-type="Python SDK" />
+</project>
\ No newline at end of file
Index: bback_car/avoidobstacle.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/bback_car/avoidobstacle.py b/bback_car/avoidobstacle.py
new file mode 100644
--- /dev/null	(date 1603950883874)
+++ b/bback_car/avoidobstacle.py	(date 1603950883874)
@@ -0,0 +1,139 @@
+import rclpy
+from rclpy.node import Node
+import RPi.GPIO as gpio
+import time
+import sys
+import signal
+from std_msgs.msg import Int32
+
+# This class is avoiding things but stick to the overall goal
+# If we are in an x y coordinate system in a two dimensional world
+# If goal is forward and something gets in the way we will try to go around it and
+# get back to x position in front of the obstacle
+
+class AvoidObstacle(Node):
+
+    def __init__(self):
+        super().__init__('avoid')
+        self.subscription = self.create_subscription(
+            Int32,
+            'frontdistance',
+            self.listener_callback,
+            10)
+        self.subscription  # prevent unused variable warning
+        self.pinMotorAForwards = 9
+        self.pinMotorABackwards = 10
+        self.pinMotorBForwards = 7
+        self.pinMotorBBackwards = 8
+        self.Frequency = 20
+        # How long the pin stays on each cycle, as a percent
+        self.DutyCycleA = 30
+        self.DutyCycleB = 30
+        # Setting the duty cycle to 0 means the motors will not turn
+        self.Stop = 0
+        gpio.setmode(gpio.BCM)
+        gpio.setwarnings(False)
+        gpio.setup(self.pinMotorAForwards, gpio.OUT)
+        gpio.setup(self.pinMotorABackwards, gpio.OUT)
+        gpio.setup(self.pinMotorBForwards, gpio.OUT)
+        gpio.setup(self.pinMotorBBackwards, gpio.OUT)
+        #Distance Variables
+        self.hownear = 15.0
+        self.reversetime = 0.5
+        self.turntime = 0.75
+
+        # Set the GPIO to software PWM at 'Frequency' Hertz
+        self.pwmMotorAForwards = gpio.PWM(self.pinMotorAForwards, self.Frequency)
+        self.pwmMotorABackwards = gpio.PWM(self.pinMotorABackwards, self.Frequency)
+        self.pwmMotorBForwards = gpio.PWM(self.pinMotorBForwards, self.Frequency)
+        self.pwmMotorBBackwards = gpio.PWM(self.pinMotorBBackwards, self.Frequency)
+
+        # Start the software PWM with a duty cycle of 0 (i.e. not moving)
+        self.pwmMotorAForwards.start(self.Stop)
+        self.pwmMotorABackwards.start(self.Stop)
+        self.pwmMotorBForwards.start(self.Stop)
+        self.pwmMotorBBackwards.start(self.Stop)
+
+    # Turn all motors off
+    def stopmotors(self):
+        self.pwmMotorAForwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorABackwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorBForwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorBBackwards.ChangeDutyCycle(self.Stop)
+
+
+    # Turn both motors forwards
+    def forwards(self):
+        self.pwmMotorAForwards.ChangeDutyCycle(self.DutyCycleA)
+        self.pwmMotorABackwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorBForwards.ChangeDutyCycle(self.DutyCycleB)
+        self.pwmMotorBBackwards.ChangeDutyCycle(self.Stop)
+
+
+    # Turn both motors backwards
+    def backwards(self):
+        self.pwmMotorAForwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorABackwards.ChangeDutyCycle(self.DutyCycleA)
+        self.pwmMotorBForwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorBBackwards.ChangeDutyCycle(self.DutyCycleB)
+
+
+    # Turn left
+    def left(self):
+        self.pwmMotorAForwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorABackwards.ChangeDutyCycle(self.DutyCycleA)
+        self.pwmMotorBForwards.ChangeDutyCycle(self.DutyCycleB)
+        self.pwmMotorBBackwards.ChangeDutyCycle(self.Stop)
+
+
+    # Turn Right
+    def right(self):
+        self.pwmMotorAForwards.ChangeDutyCycle(self.DutyCycleA)
+        self.pwmMotorABackwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorBForwards.ChangeDutyCycle(self.Stop)
+        self.pwmMotorBBackwards.ChangeDutyCycle(self.DutyCycleB)
+
+
+
+    # Move back a little, then turn right
+    def avoidobstacle(self):
+        # Back off a little
+        print("Backwards")
+        self.backwards()
+        time.sleep(self.reversetime)
+        self.stopmotors()
+
+        # Turn right
+        print("Right")
+        self.right()
+        time.sleep(self.turntime)
+        self.stopmotors()
+        time.sleep(self.turntime)
+
+    def listener_callback(self, msg):
+        self.get_logger().info('I heard: "%d"' % msg.data)
+        if msg.data < self.hownear:
+            self.stopmotors()
+            self.avoidobstacle()
+            self.forwards()
+        else:
+            self.forwards()
+
+
+
+def main(args=None):
+    rclpy.init(args=args)
+
+    avoid = Avoid()
+
+    rclpy.spin(avoid)
+
+    # Destroy the node explicitly
+    # (optional - otherwise it will be done automatically
+    # when the garbage collector destroys the node object)
+    avoid.destroy_node()
+    rclpy.shutdown()
+
+
+if __name__ == '__main__':
+    main()
Index: .idea/inspectionProfiles/profiles_settings.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/inspectionProfiles/profiles_settings.xml b/.idea/inspectionProfiles/profiles_settings.xml
new file mode 100644
--- /dev/null	(date 1603656075960)
+++ b/.idea/inspectionProfiles/profiles_settings.xml	(date 1603656075960)
@@ -0,0 +1,6 @@
+<component name="InspectionProjectProfileManager">
+  <settings>
+    <option name="USE_PROJECT_PROFILE" value="false" />
+    <version value="1.0" />
+  </settings>
+</component>
\ No newline at end of file
Index: bback_car/adrive.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/bback_car/adrive.cpp b/bback_car/adrive.cpp
new file mode 100644
--- /dev/null	(date 1603656207313)
+++ b/bback_car/adrive.cpp	(date 1603656207313)
@@ -0,0 +1,297 @@
+/*
+  :Version 2.0
+  :Author: Dragos Calin
+  :Email: dragos@intorobotics.com
+  :License: BSD
+  :Date: 20/04/2020
+  :Last update: dd/mm/YYYY
+*/
+#include <NewPing.h>
+#include <SimpleKalmanFilter.h>
+
+#define SONAR_NUM 3          //The number of sensors.
+#define MAX_DISTANCE 200     //Mad distance to detect obstacles.
+#define PING_INTERVAL 33     //Looping the pings after 33 microseconds.
+
+int LOOPING              = 10; //Loop for every 10 milliseconds.
+int DECREESE_SPEED_LOOP  = 400;//Give some time to sensors for few more readings.
+int MOVE_TO_NEW_POSITION = 500;//Wait for the new position.
+
+unsigned long _timerStart         = 0;
+unsigned long _timerStartReady    = 0;
+unsigned long _timerStartPosition = 0;
+
+uint8_t MIN_RANGE_OBSTACLE = 5; //Between 0 and 5 cm is the blind zone of the sensor.
+uint8_t MAX_RANGE_OBSTACLE = 75; //The maximum range to check if obstacle exists.
+
+uint8_t oldSensorReading[3];    //Store last valid value of the sensors.
+
+uint8_t leftSensor;             //Store the sensor's value.
+uint8_t centerSensor;
+uint8_t rightSensor;
+
+bool isObstacleLeft;           //If obstacle detected or not.
+bool isObstacleCenter;
+bool isObstacleRight;
+
+uint8_t maximumSpeed = 255; //PWM value for maximum speed.
+uint8_t minSpeed = 100; //PWM value for minimum speed.
+
+unsigned long pingTimer[SONAR_NUM]; // Holds the times when the next ping should happen for each sensor.
+unsigned int cm[SONAR_NUM];         // Where the ping distances are stored.
+uint8_t currentSensor = 0;          // Keeps track of which sensor is active.
+
+NewPing sonar[SONAR_NUM] = {
+  NewPing(3, 2, MAX_DISTANCE), // Trigger pin, echo pin, and max distance to ping.
+  NewPing(5, 4, MAX_DISTANCE),
+  NewPing(7, 6, MAX_DISTANCE)
+};
+
+SimpleKalmanFilter KF_Left(2, 2, 0.01);
+SimpleKalmanFilter KF_Center(2, 2, 0.01);
+SimpleKalmanFilter KF_Right(2, 2, 0.01);
+
+enum NavigationStates {
+  CHECK_ALL,
+  MAX_SPEED,
+  SPEED_DECREASE,
+  CHECK_OBSTACLE_POSITION,
+  LEFT,
+  CENTER,
+  RIGHT,
+  BACK
+};
+NavigationStates _navState = CHECK_ALL;
+
+//L298N motor driver pins
+byte enA = 9;
+byte in1 = 8;
+byte in2 = 11;
+byte enB = 10;
+byte in3 = 12;
+byte in4 = 13;
+
+void startTimer() {
+  _timerStart = millis();
+}
+
+void startTimerReady() {
+  _timerStartReady = millis();
+}
+
+void startTimerPosition() {
+  _timerStartPosition = millis();
+}
+
+bool isTimeForLoop(int _mSec) {
+  return (millis() - _timerStart) > _mSec;
+}
+
+bool isTimerReady(int _mSec) {
+  return (millis() - _timerStartReady) > _mSec;
+}
+
+bool isTimerPosition(int _mSec) {
+  return (millis() - _timerStartPosition) > _mSec;
+}
+
+//looping the sensors
+void sensorCycle() {
+  for (uint8_t i = 0; i < SONAR_NUM; i++) {
+    if (millis() >= pingTimer[i]) {
+      pingTimer[i] += PING_INTERVAL * SONAR_NUM;
+      if (i == 0 && currentSensor == SONAR_NUM - 1) oneSensorCycle();
+      sonar[currentSensor].timer_stop();
+      currentSensor = i;
+      cm[currentSensor] = 0;
+      sonar[currentSensor].ping_timer(echoCheck);
+    }
+  }
+}
+
+// If ping received, set the sensor distance to array.
+void echoCheck() {
+  if (sonar[currentSensor].check_timer())
+    cm[currentSensor] = sonar[currentSensor].ping_result / US_ROUNDTRIP_CM;
+}
+
+//Return the last valid value from the sensor.
+void oneSensorCycle() {
+  leftSensor   = returnLastValidRead(0, cm[0]);
+  centerSensor = returnLastValidRead(1, cm[1]);
+  rightSensor  = returnLastValidRead(2, cm[2]);
+}
+
+//If sensor value is 0, then return the last stored value different than 0.
+int returnLastValidRead(uint8_t sensorArray, uint8_t cm) {
+  if (cm != 0) {
+    return oldSensorReading[sensorArray] = cm;
+  } else {
+    return oldSensorReading[sensorArray];
+  }
+}
+
+//Apply Kalman Filter to sensor reading.
+void applyKF() {
+  isObstacleLeft = obstacleDetection(KF_Left.updateEstimate(leftSensor));
+  isObstacleCenter = obstacleDetection(KF_Center.updateEstimate(centerSensor));
+  isObstacleRight = obstacleDetection(KF_Right.updateEstimate(rightSensor));
+}
+
+//Define the minimum and maximum range of the sensors, and return true if an obstacle is in range.
+bool obstacleDetection(int sensorRange) {
+  if ((MIN_RANGE_OBSTACLE <= sensorRange) && (sensorRange <= MAX_RANGE_OBSTACLE)) return true; else return false;
+}
+
+//Obstacle avoidance algorithm.
+void obstacleAvoidance()
+{
+  switch (_navState) {
+    case CHECK_ALL: { //if no obstacle, go forward at maximum speed
+        if (isObstacleLeft == 0 && isObstacleCenter == 0 && isObstacleRight == 0) {
+          _navState = MAX_SPEED;
+        } else {
+          startTimerReady();
+          _navState = SPEED_DECREASE;
+        }
+      } break;
+
+    case MAX_SPEED: {
+        moveForward(maximumSpeed);
+        _navState = CHECK_ALL;
+      } break;
+
+    case SPEED_DECREASE: {
+        moveForward(minSpeed);
+        //Wait for few more readings at low speed and then go to check the obstacle position
+        if (isTimerReady(DECREESE_SPEED_LOOP)) _navState = CHECK_OBSTACLE_POSITION;
+      } break;
+
+    case CHECK_OBSTACLE_POSITION: {
+        //If the path is free, go again to MAX_SPEED else check the obstacle position
+        if (isObstacleLeft == 0 && isObstacleCenter == 0 && isObstacleRight == 0) {
+          _navState = MAX_SPEED;
+        }
+        else if (isObstacleLeft == 1 && isObstacleCenter == 0 && isObstacleRight == 0) {
+          startTimerPosition();
+          _navState = LEFT;
+        }
+        else if (isObstacleLeft == 0 && isObstacleCenter == 1 && isObstacleRight == 0) {
+          startTimerPosition();
+          _navState = CENTER;
+        }
+        else if (isObstacleLeft == 0 && isObstacleCenter == 0 && isObstacleRight  == 1) {
+          startTimerPosition();
+          _navState = RIGHT;
+        }
+        else if (isObstacleLeft == 1 && isObstacleCenter == 1 && isObstacleRight == 1) {
+          startTimerPosition();
+          _navState = BACK;
+        }
+      } break;
+
+    case LEFT: { //Move left and check obstacle. If obstacle exists, go again to left, else exit
+        moveLeft(minSpeed);
+        if (isTimerPosition(MOVE_TO_NEW_POSITION)) {
+          if (isObstacleLeft == 1) _navState = LEFT;
+          else _navState = CHECK_ALL;
+        }
+      } break;
+
+    case CENTER: { //If obstacle exists, go left or right
+        if (randomMove() == 1)  _navState = LEFT; else  _navState = RIGHT;
+      } break;
+
+    case RIGHT: {
+        moveRight(minSpeed);
+        if (isTimerPosition(MOVE_TO_NEW_POSITION)) {
+          if (isObstacleRight == 1) _navState = RIGHT;
+          else _navState = CHECK_ALL;
+        }
+      } break;
+
+    case BACK: {
+        moveBackward(minSpeed);
+        if (isTimerPosition(MOVE_TO_NEW_POSITION)) {
+          if (randomMove() == 1)  _navState = LEFT; else  _navState = RIGHT;
+        }
+      } break;
+  }
+}
+
+//L298N Motor Driver.
+void stopMotors() {
+  digitalWrite(in1, LOW);
+  digitalWrite(in2, LOW);
+  digitalWrite(in3, LOW);
+  digitalWrite(in4, LOW);
+}
+
+void moveForward(uint8_t pwmValue) {
+  digitalWrite(in1, HIGH);
+  digitalWrite(in2, LOW);
+  digitalWrite(in3, HIGH);
+  digitalWrite(in4, LOW);
+  analogWrite(enA, pwmValue);
+  analogWrite(enB, pwmValue);
+}
+
+void moveBackward(uint8_t pwmValue) {
+  digitalWrite(in1, LOW);
+  digitalWrite(in2, HIGH);
+  digitalWrite(in3, LOW);
+  digitalWrite(in4, HIGH);
+  analogWrite(enA, pwmValue);
+  analogWrite(enB, pwmValue);
+}
+
+void moveLeft(uint8_t pwmValue) {
+  digitalWrite(in1, LOW); //Left wheel backward.
+  digitalWrite(in2, HIGH);
+  digitalWrite(in3, HIGH);//Right wheel forward.
+  digitalWrite(in4, LOW);
+  analogWrite(enA, pwmValue);
+  analogWrite(enB, pwmValue);
+}
+
+void moveRight(uint8_t pwmValue) {
+  digitalWrite(in1, HIGH); //Left wheel forward.
+  digitalWrite(in2, LOW);
+  digitalWrite(in3, LOW);//Right wheel backward.
+  digitalWrite(in4, HIGH);
+  analogWrite(enA, pwmValue);
+  analogWrite(enB, pwmValue);
+}
+
+//Return 1 or 0 from a random number.
+int randomMove() {
+  uint8_t rnd_number = random(1, 100);
+  return rnd_number % 2;
+
+}
+/*SETUP & LOOP*/
+void setup() {
+  Serial.begin(115200);
+
+  pingTimer[0] = millis() + 75;
+  for (uint8_t i = 1; i < SONAR_NUM; i++)
+    pingTimer[i] = pingTimer[i - 1] + PING_INTERVAL;
+
+  pinMode(enA, OUTPUT);
+  pinMode(in1, OUTPUT);
+  pinMode(in2, OUTPUT);
+  pinMode(enB, OUTPUT);
+  pinMode(in3, OUTPUT);
+  pinMode(in4, OUTPUT);
+
+  stopMotors();
+}
+
+void loop() {
+  if (isTimeForLoop(LOOPING)) {
+    sensorCycle();
+    applyKF();
+    obstacleAvoidance();
+    startTimer();
+  }
+}
Index: .idea/.gitignore
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/.gitignore b/.idea/.gitignore
new file mode 100644
--- /dev/null	(date 1603656109365)
+++ b/.idea/.gitignore	(date 1603656109365)
@@ -0,0 +1,3 @@
+# Default ignored files
+/shelf/
+/workspace.xml
Index: .idea/modules.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/modules.xml b/.idea/modules.xml
new file mode 100644
--- /dev/null	(date 1603656075952)
+++ b/.idea/modules.xml	(date 1603656075952)
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="ProjectModuleManager">
+    <modules>
+      <module fileurl="file://$PROJECT_DIR$/.idea/robot_car2.iml" filepath="$PROJECT_DIR$/.idea/robot_car2.iml" />
+    </modules>
+  </component>
+</project>
\ No newline at end of file
Index: .idea/vcs.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/.idea/vcs.xml b/.idea/vcs.xml
new file mode 100644
--- /dev/null	(date 1603656075964)
+++ b/.idea/vcs.xml	(date 1603656075964)
@@ -0,0 +1,6 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project version="4">
+  <component name="VcsDirectoryMappings">
+    <mapping directory="$PROJECT_DIR$" vcs="Git" />
+  </component>
+</project>
\ No newline at end of file
